pages/typeahead.tsx

```diff
@@ -0,0 +1,171 @@
+import React, { EventHandler, KeyboardEventHandler, useRef, useState } from "react"
+import { User } from "../types/types";
+
+const usersList: User[] = [
+    {
+        name: "Jayz",
+    },
+    {
+        name: "Ash",
+    },
+    {
+        name: "Deep"
+    },
+    {
+        name: "Yelena",
+    },
+    {
+        name: "Eren"
+    },
+    {
+        name: "Mikasa"
+    },
+    {
+        name: "Armin"
+    },
+    {
+        name: "JayzTheGoat"
+    },
+    {
+        name: "Ashwin"
+    },
+    {
+        name: "DeepXAshwin"
+    }
+];
+
+const typeahead = () => {
+    const [users, setUsers] = useState<User[]>(usersList);
+    const [isDropdownOpen, setIsDropdownOpen] = useState<boolean>(false);
+    const [selectedUsers, setSelectedUsers] = useState<User[]>([]);
+    const [typeaheadValue, setTypeaheadValue] = useState<string>("");
+    const textInput = useRef<HTMLInputElement>(null);
+    // TODO: Either use tracked "users" or use tracked "displayedUsers"
+    const [displayedUsers, setDisplayedUsers] = useState<User[]>(users);
+
+    const onInputClick = (event: React.MouseEvent<HTMLInputElement>) => {
+        toggleSuggestionsDropdown();
+        // if (isDropdownOpen) {
+        //     document.querySelector('.suggestions-box')?.focus();
+        // }
+    }
+
+    const toggleSuggestionsDropdown = () => {
+        setIsDropdownOpen((isDropdownOpen) => !isDropdownOpen);
+    }
+
+    const discardLastSelectedUser = () => {
+        const lastSelectedUser  = selectedUsers && selectedUsers[-1];
+        setDisplayedUsers((users) => [...users, lastSelectedUser]);
+        setSelectedUsers((users) => users.filter(user => user.name !== lastSelectedUser.name));
+    }
+
+    const onSelectListItem = (selectedUser: User) => {
+        setDisplayedUsers((users) => users.filter(user => user.name !== selectedUser.name));
+        setSelectedUsers((users) => [...users, selectedUser]);
+    }
+
+    const onDeselectItem = (selectedUser: User) => {
+        setDisplayedUsers((users) => [...users, selectedUser]);
+        setSelectedUsers((users) => users.filter(user => user.name !== selectedUser.name));
+    }
+
+    const getUserswithMatchingName = (inputValue: string): User[] => {
+        return users.filter((user) => (user.name.slice(0, inputValue.length) === inputValue)) || [];
+    }
+
+    const onTypeaheadInput = (event: React.ChangeEvent) => {
+        const textInputValue = (event.target as HTMLInputElement).value;
+        setTypeaheadValue(textInputValue);
+        setDisplayedUsers(getUserswithMatchingName(textInputValue));
+    }
+
+    const onKeyUpInput = (event: React.KeyboardEvent) => {
+        event.preventDefault();
+        if (event.key === "Enter") {
+            toggleSuggestionsDropdown();
+        }
+        if (!textInput.current.value && event.key === "Backspace") {
+            const selected = document.querySelector(".selected-users-container .selected-user-item:last-of-type");
```

Comment:
dont do this ever

Reply:
@ashwinkhode Thanks for review, will update to use Refs here.